### PR TITLE
`get_default_shuffle_method` raises if `pyarrow` is outdated

### PR DESCRIPTION
**Pull request by hendrikmakait**
- [x] Sibling to and blocked by dask/distributed#8157
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by fjetter — APPROVED** [review]

**Comment by jrbourbeau** [comment]
Marking as `DNM` for now since the `distributed` PR needs to go in first

@hendrikmakait we're getting a `FAILED dask/tests/test_distributed.py::test_bag_groupby_default - RuntimeError: P2P shuffling requires pyarrow>=7.0.0` failure here now. Maybe this will go away once the `distributed` PR is merged (I've not looked into it at all). Just wanted to double check whether additional changes are still needed or not

**Comment by hendrikmakait** [comment]
@jrbourbeau: I checked `test_distributed.py` locally and the following tests will start failing because the installed `pyarrow` version does not match the minimum requirement of `p2p`:

* `test_fused_blockwise_dataframe_merge`
* `test_map_partitions_df_input`

There are two ways to fix this:

1. Increase the installed version to `pyarrow>=12.0.0`
2. Force the tests to use task-based shuffling

From what I understand, we _want_ to test those lower version, so I'll take a quick stab at (2.).

File: continuous_integration/environment-3.11.yaml
Comment (hendrikmakait) [inline continuous_integration/environment-3.11.yaml:78]: REVERT me and my siblings

File: dask/utils.py
Comment (hendrikmakait) [inline dask/utils.py:2126]: `pyarrow` is not installed, but we don't care.

File: dask/bag/core.py
Comment (hendrikmakait) [inline dask/bag/core.py:1531]: `pyarrow` does not match `p2p` requirements, but we don't care.

**Comment by hendrikmakait** [comment]
When using the sibling branch on CI, tests are green: https://github.com/dask/dask/pull/10496/commits/f9a11b29d5a5dbf4b00e6520bd5a629abe4d18c1

File: dask/bag/core.py
Comment (jrbourbeau) [reply dask/bag/core.py:1531]: Wait, I'm confused. I wouldn't expect `get_default_shuffle_method` to ever raise

Comment (hendrikmakait) [reply dask/bag/core.py:1531]: See https://github.com/dask/distributed/pull/8157#issuecomment-1709648465

The idea is that we now require a pretty recent version of `pyarrow` so instead of silently falling back to `tasks` (or requiring it for all of `distributed`), we fail to alert users that have `pyarrow` installed but not the minimum version. Please defer to @fjetter to hash this out, he's taking over the PR while I'm out.

Comment (hendrikmakait) [reply dask/bag/core.py:1531]: The more troublesome issue is that `get_default_shuffle_method` gets shared between `dataframe` and `bag`.

Comment (fjetter) [reply dask/bag/core.py:1531]: @jrbourbeau is this a problem for you? I want to avoid users accidentally falling back to tasks if they have an old pyarrow installed. This exception is raised early and they can either fix it by upgrading or by setting the config of shuffle to tasks.

Comment (fjetter) [reply dask/bag/core.py:1531]: @jrbourbeau I will move forward with this now since you didn't reply in a little over a week. I'd like this to sit in main for at least a day before we release

**Comment by fjetter** [comment]
merged main again. This should also show that the changes are compatible with dask/distributed after merging https://github.com/dask/distributed/pull/8157

Once build is done (and green) we can merge

**Comment by fjetter** [comment]
There are plenty of failures on gpu CI but I think this is unrelated cc @dask/gpu

File: dask/bag/core.py
Comment (jrbourbeau) [reply dask/bag/core.py:1531]: Sounds good if CI is happy

**Comment by fjetter** [comment]
I will move forward with merging this since otherwise the dask/distributed repo might also break. If the gpuCI failure is related, we will have to investigate this more closely and possible revert the distributed PR

**Pull request merged**
